torch/csrc/utils/future.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <ATen/core/ivalue.h>
+
+namespace torch {
+
+namespace utils {
+
+// FutuerError inherits from std::exception, it can return const char* or
+// std::string error message
+struct TORCH_API FutureError final : public std::exception {
+  FutureError(std::string&& error_msg_) : error_msg(std::move(error_msg_)) {}
+
+  FutureError() = default;
+
+  const char* what() const noexcept override {
+    return error_msg.c_str();
+  }
+
+  std::string errMsg() const {
+    return error_msg;
+  }
+
+  std::string error_msg;
+};
+
+// This class holds a value of type T that will be ready in the future.
+// Most implementation is copied from FutureMessage and
+// c10::ivalue::Future
+template <typename T>
+class TORCH_API Future final {
+ public:
+  using Callback = std::function<void(const T&, bool, const FutureError&)>;
```

Comment:
Shall we split code copy and FutureError into separate PRs? Or is there any reason that we have do them in one?

Reply:
wondering: maybe a template like:
    void(const T&., const FutureError* err)

where err != nullptr -> present.
or maybe two pointers? i.e. void(const T*, const FutureError *)

Reply:
since this is generic Future<T>,  I'm copying from both FutureMessage and c10::ivalue::Future here

Reply:
right, this is suggested by @pietern as well, I'm killing the flag